tests/xmap_test.py

```diff
@@ -335,6 +532,47 @@ def f(x, y):
 
     self.assertAllClose(z, jnp.einsum('nij,njk->nik', x, y))
 
+  @parameterized.named_parameters(jtu.cases_from_list(
```

Comment:
I am curious how many raw test cases are here? If there are many, then we pay the cost to generate them and then we throw most of them out. When testing jax2tf with this kind of combinatorial test case generation I have run several times into cases when the interesting use cases where systematically ignored.  An alternative strategy is to generate a sequence of families of test cases, each with shallow-nested for-loops.

However, I do not understand the code being tested, and how many interesting paths through it there are. Perhaps there are really lots of paths.

Reply:
That's a good point to raise! It seems the relevant figure of merit is test generation time. On my machine it takes less than 1.5 seconds to generate all the test cases here (computed by comparing pytest's reported time running tests to timing the pytest command itself from my shell).

Each test case here is a valid pdot call and a valid xmap scheduling for that call, and there are no duplicates. Moreover this seems like the place where we most want exhaustive tests: any of these calls could reasonably appear in the wild, and we want xmap + pdot to be rock solid since we expect them to get a lot of use.

Maybe we can revisit if the test generation time becomes unreasonable, like tens of seconds?

Reply:
It's a common theme in many of our tests. There's the same problem in other pmap and xmap tests. I've been briefly entertaining an idea of switching the parametrization from generating a full list only to downsample it, to taking a method that generates a single case, parametrized by some PRNG/oracle that determines which choices to make. So instead of saying
```py
jtu.cases_form_list(
  {"x": x, "y": y, "z": z}
  for x in range(10000000)
  for y in range(10000000)
  for z in range(10000000))
```
we would instead say
```py
jtu.cases_from_sampler(
  lambda sampler: {"x": sampler.sample(10000000),
                   "y": sampler.sample(10000000),
                   "z": sampler.sample(10000000)})
```
Technically the latter could still be used to generate an exhaustive list of tests, but we also would have the option to only evaluate its body a few times to generate a small subset of the space of potential test space.

Reply:
1.5s is not a lot if we have a single huge parametrized test, but I expect us to have more of those in the future. And waiting for 20s just to generate a subset of test cases seems unreasonable.

Reply:
Besides the testing slowdown, I had run into another problem with such tests. If this pattern generates 1000+ test cases from which we sample 10 by default or 25 in CI (and we pick always the same ones), it is quite possible that some important use cases are never tested until somebody runs the test manually with a very large NUM_GENERATED_CASES. That is in part why I asked how many tests are generated here. It looks likely to be in the thousands, and if it takes more than 1sec to generate them then it is possibly many thousands.

Reply:
One of my parametrizations generates ~250k cases and finishes in ~1.5s, so I'd expect the number to be O(100k) here too

Reply:
> it is quite possible that some important use cases are never tested until somebody runs the test manually with a very large NUM_GENERATED_CASES

That seems like a pretty good process to me. It's worked well for us in the past, e.g. for the lazy sublanguage and for our early development of the more complex lax primitives (like conv_general_dilated, dot_general, etc). During development (or revision) we manually run lots of them, then we rely on a subset for CI for regressions. (Regressions seem less likely in these cases.)

We took an alternative route for einsum testing, where we just added cases that we thought of. It didn't work well: the result was we missed lots of edge cases that had to be cleaned up as users ran into them! The space of einsum configurations was just too tricky to cover thoroughly with a priori thinking. I concluded systematic tests can be really useful.

Luckily we don't have to choose between systematic coverage and tests for specific edge cases: we can do both! Perhaps a useful pattern (maybe even one that can be automated?) is to add explicit cases whenever we see a failure in the manually-run exhaustive tests. That way, we could have the exhaustive tests to run manually when we want to (during development or revision to the code in question), but perhaps by default their generation is skipped for speed (i.e. CI could skip generating them entirely, or we could figure out a more efficient way to generate a subsample). At the same time we can also have good explicit edge case coverage, both from whatever hand-crafting we come up with and from saving past failures of our systematic search.

WDYT?

Reply:
By the way, as per my earlier comment in this thread, I think the best option here is to keep figuring out a faster approach to (subsampled) systematic testing while also not blocking this particular PR on it, since this PR doesn't itself introduce any problematic test generation times. That is, we should solve the test generation timing problem before it becomes an issue with these xmap tests, but it doesn't need to be solved in this PR.

Reply:
It seems there are 15295 tests generated in my current file, though I've been changing things so it may not reflect the same number as an earlier version. I did a more accurate timing measure and found it required took less than 0.09 seconds to generate them, and 0.96 seconds to generate and subsample them. So we're okay on timing for now I think!

Reply:
I like the idea of keeping the option to run exhaustive tests, and also manually adding cases that we think are important to cover. Without the latter, we may not spot regressions until users report them, and then we manually run with many generated cases. I, for one, do not have the habit to routinely run with many generated cases, nor it is on my mind to ask contributors to do so when I review their code. Perhaps we can change this habit.

Reply:
(Now it's at 0.3 seconds to generate and subsample them, based on generating more possible schedules as per Adam's comment below.)